zstd.cc

```diff
@@ -22,15 +22,41 @@ static const sstring COMPRESSOR_NAME = compressor::namespace_prefix + "ZstdCompr
 class zstd_processor : public compressor {
     int _compression_level = 3;
 
-    // Manages memory for the compression context.
-    std::unique_ptr<char[], free_deleter> _cctx_raw;
-    // Compression context. Observer of _cctx_raw.
-    ZSTD_CCtx* _cctx;
-
-    // Manages memory for the decompression context.
-    std::unique_ptr<char[], free_deleter> _dctx_raw;
-    // Decompression context. Observer of _dctx_raw.
-    ZSTD_DCtx* _dctx;
+    // We share buffers among instances of the compressor to limit the number
+    // of large allocations and to prevent OOM in case there are many
+    // compressors alive simultaneously.
+    //
+    // However, we don't share buffers bigger than a certain arbitrary
+    // threshold. The threshold is expected to be bigger than any buffer
+    // encountered with reasonable settings.
+    //
+    // This is a simple protection from huge shared leftover compression
+    // buffers staying in memory indefinitely. Huge buffers are not expected to
+    // arise during normal operation, but if a user messes up his table
+    // compression settings by setting 64 MiB chunks and max compression level,
+    // the compression context can grow up to 680 MiB. If we share this buffer,
+    // it will live and take up the memory even after the user fixes the
+    // settings.
+    //
+    // The max context size for the default compression level 3 is 1.3 MiB.
+    // The max context size for max compression level 22 and chunks of size
+    // 64 KiB is 1.8 MiB. So 2 MiB seems like a good threshold for "reasonable"
+    // setups.
```

Comment:
An alternative approach is to decay the buffer sizes over time. I don't see an advantage to one over the other, 2 MiB/shard is reasonable overhead and I agree exceeding it will be rare (if traumatic).